<launch>
  <!-- Arguments -->
    <arg name="node_id" default="$(env ROBOT_NAME)"/>
    <arg name="node_type" default="$(env ROBOT_TYPE)"/>
    <arg name="endpoint" default="$(env MQTT_HOST)"/>
    <arg name="port" default="$(env MQTT_PORT)"/>
    <arg name="base_directory" default="$(find multirobot_sim)"/>
    <arg name="secret" default="$(env SECRET)"/>
    <arg name="update_interval" default="$(env UPDATE_INTERVAL)"/>
    <arg name="max_delay" default="10.0"/>
    <arg name="schema" default="$(arg base_directory)/schema/schema.sql"/>
    <arg name="timeout_interval" default="10.0"/>
    <arg name="min_nodes_num" default="$(env MIN_NODES_NUM)"/>

  <!--Robot namespace-->
    <group ns="$(arg node_id)">
        <!-- Load the controller -->
        <node pkg="multirobot_sim" type="roschain.py" name="roschain" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="node_type" value="$(arg node_type)"/>
        </node>
        <node pkg="multirobot_sim" type="consensus.py" name="consensus" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="node_type" value="$(arg node_type)"/>
            <param name="timeout_interval" value="$(arg timeout_interval)"/>
        </node>   
        <node pkg="multirobot_sim" type="network.py" name="network" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="node_type" value="$(arg node_type)"/>
        </node>   
        <node pkg="multirobot_sim" type="discovery.py" name="discovery" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="node_type" value="$(arg node_type)"/>
            <param name="secret" value="$(arg secret)"/>
            <param name="max_delay" value="$(arg max_delay)"/>
        </node>   
        <node pkg="multirobot_sim" type="heartbeat.py" name="heartbeat" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="node_type" value="$(arg node_type)"/>
            <param name="max_delay" value="$(arg max_delay)"/>
        </node>   
        <node pkg="multirobot_sim" type="key_store.py" name="key_store" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="public_key_file" value="$(arg base_directory)/files/pk.pem"/>
            <param name="private_key_file" value="$(arg base_directory)/files/sk.pem"/>
        </node>   
        <node pkg="multirobot_sim" type="connector.py" name="connector" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="endpoint" value="$(arg endpoint)"/>
            <param name="port" value="$(arg port)"/>
        </node>  
        <node pkg="multirobot_sim" type="blockchain.py" name="blockchain" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="node_type" value="$(arg node_type)"/>
            <param name="base_dir" value="$(arg base_directory)/files"/>
            <param name="secret" value="$(arg secret)"/>
        </node>  
        <node pkg="multirobot_sim" type="session.py" name="session" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
        </node> 
        <node pkg="multirobot_sim" type="database.py" name="database" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="db_dir" value="$(arg base_directory)/files"/>
            <param name="schema" value="$(arg schema)"/>
        </node>  
        <node pkg="multirobot_sim" type="dummy_transactions.py" name="dummy_transactions" output="screen" required="true">
            <param name="node_id" value="$(arg node_id)"/>
            <param name="node_type" value="$(arg node_type)"/>
            <param name="update_interval" value="$(arg update_interval)"/>
        </node>   
    </group>
</launch>
